proposals/nnnn-rpc-design-refactoring.md

```diff
@@ -0,0 +1,150 @@
+# RPC design refactoring
+
+* Proposal: [SDL-NNNN](nnnn-rpc-design-refactoring.md)
+* Author: [Alexander Kutsan](https://github.com/LuxoftAKutsan)
+* Status: **Awaiting review**
+* Impacted Platforms: [Core]
+
+## Introduction
+
+This proposal is to rework the design of the SDL core application manager layer in order to:
+ - Reduce coupling between components
+ - Reduce responsibility of the ApplicationManager class
+ - Provide a plugin architecture for adding additional functionality and RPC groups easily.
+
+## Motivation
+
+The SDL core project is actively implementing new features and new functionality.  
+To keep the project maintainable, on track and to reduce risks of regression during implementation
+refactoring of application manager layer is necessary.  
+Currently, extending SDL functionality may require code duplication. 
+
+Currently, extending SDL functionality may require code duplication. 
+The current RC module duplicates logic of the application_manager layer -- there is separate request controller, separate mechanism to call policy checks, and RPC's processing. Currently any common logic that is implemented in the application_manager layer needs to be duplicated in request controller -- this results in bugs and redundant code; it also requires more time for implementation.
+
+The following proposed solution will provide for better management and extensibility of SDL core RC components, without duplicating SDL core code:
```

Comment:
The following proposed solution will provide for better management and extensibility of SDL core RC components, without duplicating SDL core code:
should be changed to:
The following proposed solution will provide for better management and extensibility of SDL core and RC components, without duplicating SDL core code:

Reply:
@mrapitis done in 7bb2896639e1d74563efc7ff368674f9894b3f0c